plutus-playground-client/src/Chain/View.purs

```diff
@@ -0,0 +1,292 @@
+triangleRight = div [ class_ $ ClassName "triangle-right" ] []
+
+feeView :: forall p i. Ada -> HTML p i
+feeView (Lovelace { getLovelace: 0 }) = text ""
+
+feeView txFee =
+  div [ classes [ card, entry, ClassName "fee" ] ]
+    [ cardHeader_ [ triangleRight, text "Fee" ]
+    , cardBody_
+        [ div_ [ code_ [ text $ show txFee ] ]
+        ]
+    ]
+
+forgeView :: forall p i. Value -> HTML p i
+forgeView (Value { getValue: (AssocMap.Map []) }) = text ""
+
+forgeView txForge =
+  div [ classes [ card, entry, ClassName "forge" ] ]
+    [ cardHeader_ [ triangleRight, text "Forge" ]
+    , cardBody_
+        [ valueView txForge
+        ]
+    ]
+
+balancesView :: forall p i. SequenceId -> Map PubKey Wallet -> Map TxOutType Value -> HTML p i
+balancesView sequenceId walletKeys m =
+  div []
+    [ h2_
+        [ text "Closing Balances"
+        , nbsp
+        , small_
+            [ text "(as at "
+            , showSequenceId sequenceId
+            , text ")"
+            ]
+        ]
+    , row_ (balanceView <$> Map.toUnfoldable m)
+    ]
+  where
+  balanceView :: TxOutType /\ Value -> HTML p i
+  balanceView (txOutType /\ value) =
+    col3_
+      [ div
+          [ classes [ card, entry, outTypeClass txOutType ]
+          ]
+          [ cardHeaderTxOutTypeView false walletKeys txOutType
+          , cardBody_ [ valueView value ]
+          ]
+      ]
+
+showSequenceId :: forall p i. SequenceId -> HTML p i
+showSequenceId (SequenceId { slotIndex, txIndex }) =
+  span_
+    [ text "Slot #"
+    , text $ show slotIndex
+    , text ", "
+    , text "Transaction #"
+    , text $ show txIndex
+    ]
+
+txOutOfView :: forall p. Boolean -> Map PubKey Wallet -> TxOutOf String -> HTML p (Query Unit)
+txOutOfView showArrow walletKeys txOutOf@(TxOutOf { txOutAddress, txOutType, txOutValue }) =
+  div
+    [ classes [ card, entry, outTypeClass txOutType ]
+    , onClick
+        $ const
+        $ Just
+        $ action
+        $ SetChainFocus
+        $ Just
+        $ FocusAddress txOutAddress
+    ]
+    [ cardHeaderTxOutTypeView showArrow walletKeys txOutType
+    , cardBody_
+        [ valueView txOutValue ]
+    ]
+
+outTypeClass :: TxOutType -> ClassName
+outTypeClass (PayToPubKey _) = ClassName "wallet"
+
+outTypeClass (PayToScript _) = ClassName "script"
+
+cardHeaderTxOutTypeView :: forall p i. Boolean -> Map PubKey Wallet -> TxOutType -> HTML p i
+cardHeaderTxOutTypeView showArrow walletKeys txOutType =
+  div [ classes [ cardHeader, textTruncate ] ]
+    [ if showArrow then triangleRight else text ""
+    , txOutTypeView walletKeys txOutType
+    ]
+
+txOutTypeView :: forall p i. Map PubKey Wallet -> TxOutType -> HTML p i
+txOutTypeView walletKeys (PayToPubKey pubKey) = case Map.lookup pubKey walletKeys of
+  Nothing -> showPubKey pubKey
+  Just (Wallet { getWallet: n }) ->
+    span_
+      [ showPubKey pubKey
+      , br_
+      , small_
+          [ text "Wallet"
+          , nbsp
+          , text $ show n
```

Comment:
maybe I'm a bad person but I wish PS had string interpolation built in

Reply:
I'd prefer something typed.

Oh look! https://pursuit.purescript.org/packages/purescript-formatting/1.0.5

:-D